dns_server/src/crawler/mod.rs

```diff
@@ -0,0 +1,560 @@
+    fail_count: u32,
+
+    /// Whether this address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerConfig {
+    pub add_node: Vec<String>,
+
+    pub network: [u8; 4],
+
+    pub p2p_port: u16,
+}
+
+pub struct Crawler<N: NetworkingService, S> {
+    /// Crawler config
+    config: CrawlerConfig,
+
+    /// Backend's ConnectivityHandle
+    conn: N::ConnectivityHandle,
+
+    /// Backend's SyncingMessagingHandle
+    sync: N::SyncingMessagingHandle,
+
+    /// Map of all known addresses (including currently unreachable)
+    addresses: BTreeMap<N::Address, AddressData>,
+
+    /// Map of all currently connected peers
+    peers: BTreeMap<N::PeerId, N::Address>,
+
+    /// Storage implementation
+    storage: S,
+
+    /// Channel used to manage the DNS server
+    command_tx: mpsc::UnboundedSender<ServerCommands>,
+}
+
+impl<N: NetworkingService + 'static, S: DnsServerStorage> Crawler<N, S>
+where
+    N::SyncingMessagingHandle: SyncingMessagingService<N>,
+    N::ConnectivityHandle: ConnectivityService<N>,
+    DnsServerError: From<<<N as NetworkingService>::Address as FromStr>::Err>,
+{
+    pub fn new(
+        config: CrawlerConfig,
+        conn: N::ConnectivityHandle,
+        sync: N::SyncingMessagingHandle,
+        storage: S,
+        command_tx: mpsc::UnboundedSender<ServerCommands>,
+    ) -> Result<Self, DnsServerError> {
+        // Load all persistent addresses
+        let tx = storage.transaction_ro()?;
+
+        let storage_version = tx.get_version()?;
+        if storage_version.is_some() && storage_version != Some(STORAGE_VERSION) {
+            return Err(DnsServerError::Other("Unexpected storage version"));
+        }
+
+        let mut addresses = BTreeMap::new();
+        for address in tx.get_addresses()?.iter().filter_map(|address| address.parse().ok()) {
+            Self::new_address(&mut addresses, address, true);
+        }
+        tx.close();
+
+        if storage_version.is_none() {
+            let mut tx = storage.transaction_rw()?;
+            tx.set_version(STORAGE_VERSION)?;
+            tx.commit()?;
+        }
+
+        // Add addresses that were specified from the command line as reachable
+        for address in config.add_node.iter() {
+            let address = address.parse()?;
+            Self::new_address(&mut addresses, address, true);
+        }
+
+        Ok(Self {
+            config,
+            conn,
+            sync,
+            addresses,
+            peers: BTreeMap::new(),
+            storage,
+            command_tx,
+        })
+    }
+
+    fn handle_conn_request(
+        &mut self,
+        _peer_id: N::PeerId,
+        request_id: N::PeerRequestId,
+        request: PeerManagerRequest,
+    ) {
+        match request {
+            PeerManagerRequest::AddrListRequest(_) => {
+                // Ignored
```

Comment:
Why not provide an address list? I don't know if it's the right thing to do, but I'm interested in looking into this as a possibility.

Reply:
For security the nodes only load addresses (in a list) from outbound connections and the DNS server does not accept inbound connections at all. So normally nobody would do that. And the DNS server has other means to sharing addresses (via the DNS protocol).
Later we might implement loading addresses from the nodes to the server.

Reply:
What are the security concerns?

Reply:
This is because inbound peers should be treated as more dangerous than outbound peers. An attacker could easily target a specific node if it accepts inbound connections, connecting to it from different IPs, for example.

Reply:
Gotcha.